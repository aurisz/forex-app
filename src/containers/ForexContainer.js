import React, { Component } from 'react';
import { Container } from 'reactstrap';

import ForexBaseInput from '../components/ForexBaseInput';
import ForexAddCurrency from '../components/ForexAddCurrency';
import ForexDisplayList from '../components/ForexDisplayList';

import { formatCurrencyOptions } from '../utils';

class ForexContainer extends Component {
  state = {
    error: null,
    isLoaded: false,
    baseValue: 10.0,
    currencyRates: {},
    currencyDisplays: ['IDR', 'EUR', 'GBP', 'SGD'],
    selectedNewCurrency: '',
    isAddMoreCurrency: false
  };

  componentDidMount() {
    fetch('https://api.exchangeratesapi.io/latest?base=USD')
      .then(res => res.json())
      .then(
        result => {
          this.setState({
            isLoaded: true,
            currencyRates: result.rates
          });
        },
        error => {
          this.setState({
            isLoaded: true,
            error
          });
        }
      );
  }

  handleInputChange = e => {
    const { value, name } = e.target;

    // special case for base value input
    let inputValue = value;
    if (name === 'baseValue') {
      inputValue = parseFloat(value) || '';
    }

    this.setState({ [name]: inputValue });
  };

  handleClickAddMoreCurrency = () => {
    this.setState({ isAddMoreCurrency: true });
  };

  handleAddCurrency = () => {
    const { currencyDisplays, selectedNewCurrency } = this.state;

    this.setState({
      currencyDisplays: [...currencyDisplays, selectedNewCurrency],
      selectedNewCurrency: '',
      isAddMoreCurrency: false
    });
  };

  handleDeleteCurrency = deletedCurrency => {
    const { currencyDisplays } = this.state;

    // remove deleted currency from displayed currency state
    const filteredCurrency = currencyDisplays.filter(
      v => v !== deletedCurrency
    );

    this.setState({
      currencyDisplays: filteredCurrency
    });
  };

  render() {
    const {
      error,
      isLoaded,
      baseValue,
      currencyRates,
      currencyDisplays,
      selectedNewCurrency,
      isAddMoreCurrency
    } = this.state;

    return (
      <Container>
        <ForexBaseInput
          value={baseValue}
          onInputChange={this.handleInputChange}
        />

        <ForexDisplayList
          error={error}
          isLoaded={isLoaded}
          baseValue={baseValue}
          currencies={currencyDisplays}
          rates={currencyRates}
          handleDeleteCurrency={this.handleDeleteCurrency}
        />

        <ForexAddCurrency
          options={formatCurrencyOptions(currencyRates, currencyDisplays)}
          value={selectedNewCurrency}
          isAddMoreCurrency={isAddMoreCurrency}
          onInputChange={this.handleInputChange}
          onSubmitAddCurrency={this.handleAddCurrency}
          onClickAddMoreCurrency={this.handleClickAddMoreCurrency}
        />
      </Container>
    );
  }
}

export default ForexContainer;
