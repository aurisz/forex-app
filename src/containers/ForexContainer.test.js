import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';

import ForexContainer from './ForexContainer';

describe('<ForexContainer />', () => {
  it('renders without crashing', () => {
    const wrapper = shallow(<ForexContainer />);
    expect(wrapper.exists()).toBe(true);
  });
  it('renders correctly', () => {
    const tree = renderer.create(<ForexContainer />).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('should be handling input change for base value', () => {
    const wrapper = shallow(<ForexContainer />);
    wrapper
      .instance()
      .handleInputChange({ target: { value: 15, name: 'baseValue' } });
    expect(wrapper.instance().state.baseValue).toEqual(15);
  });
  it('should be handling input change for everything else', () => {
    const wrapper = shallow(<ForexContainer />);
    wrapper.instance().handleInputChange({
      target: { value: 'EUR', name: 'selectedNewCurrency' }
    });
    expect(wrapper.instance().state.selectedNewCurrency).toEqual('EUR');
  });
  it('should be handling click add more currency', () => {
    const wrapper = shallow(<ForexContainer />);
    wrapper.instance().handleClickAddMoreCurrency();
    expect(wrapper.instance().state.isAddMoreCurrency).toEqual(true);
  });
  it('should be handling add currency submit', () => {
    const wrapper = shallow(<ForexContainer />);

    wrapper.setState({ selectedNewCurrency: 'NOK' });

    wrapper.instance().handleAddCurrency();
    expect(wrapper.instance().state.currencyDisplays).toEqual([
      'IDR',
      'EUR',
      'GBP',
      'SGD',
      'NOK'
    ]);
    expect(wrapper.instance().state.selectedNewCurrency).toBe('');
    expect(wrapper.instance().state.isAddMoreCurrency).toEqual(false);
  });
  it('should be handling delete currency from list', () => {
    const wrapper = shallow(<ForexContainer />);

    wrapper.instance().handleDeleteCurrency('IDR');
    expect(wrapper.instance().state.currencyDisplays).toEqual([
      'EUR',
      'GBP',
      'SGD'
    ]);
  });
});
