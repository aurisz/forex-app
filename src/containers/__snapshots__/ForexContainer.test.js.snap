// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ForexContainer /> renders correctly 1`] = `
<div
  className="container"
>
  <div
    className="header"
  >
    <div
      className="row"
    >
      <div
        className="col"
      >
        <p>
          <em>
            USD - United States Dollars
          </em>
        </p>
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="col"
      >
        <h3>
          USD
        </h3>
      </div>
      <div
        className="col"
      >
        <input
          className="form-control"
          id="baseValueInput"
          name="baseValue"
          onChange={[Function]}
          placeholder="Enter Base Currency Value Here"
          step="any"
          type="number"
          value={10}
        />
      </div>
    </div>
  </div>
  <div
    className="loading"
  >
    Loading...
  </div>
  <div
    className="forex-add-currency"
  >
    <button
      aria-label={null}
      className="btn btn-info btn-block"
      id="addMoreCurrency"
      onClick={[Function]}
      type="button"
    >
      (+) Add More Currencies
    </button>
  </div>
</div>
`;
